carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/FeatureGenerator.java

```diff
@@ -0,0 +1,339 @@
+/*
+ *  Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.maven.p2.feature.generate;
+
+import org.apache.maven.model.Resource;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.MavenProjectHelper;
+import org.codehaus.plexus.util.DirectoryScanner;
+import org.codehaus.plexus.util.FileUtils;
+import org.wso2.maven.p2.beans.CarbonArtifact;
+import org.wso2.maven.p2.exceptions.CarbonArtifactNotFoundException;
+import org.wso2.maven.p2.exceptions.MissingRequiredPropertyException;
+import org.wso2.maven.p2.exceptions.OSGIInformationExtractionException;
+import org.wso2.maven.p2.feature.generate.utils.FeatureFileGeneratorUtils;
+import org.wso2.maven.p2.utils.BundleUtils;
+import org.wso2.maven.p2.utils.DependencyResolver;
+import org.wso2.maven.p2.utils.FileManagementUtil;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.TransformerException;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * FeatureGenerator takes parameters from the pom.xml and generates the feature.
+ */
+public class FeatureGenerator {
+
+    private final FeatureResourceBundle resourceBundle;
+    private final MavenProject project;
+    private final MavenProjectHelper projectHelper;
+
+    private File rowOutputFolder;
+    private File featureIdFolder;
+    private File pluginsFolder;
+    private File featureXmlFile;
+    private File p2InfFile;
+    private File featurePropertyFile;
+    private File featureManifestFile;
+    private File featureZipFile;
+
+    private HashMap<String, CarbonArtifact> dependentBundles;
```

Comment:
Sticking to more abstract data types would allow you better extend the functionalities later on. Therefore, instead of using concrete types such as "HashMap" you can simply make the variables use the respective abstractions.